### PR TITLE
Stress limiter rheology

### PR DESCRIPTION
**Pull request by naliboff**
Include a stress limiter rheology option in the visco_plastic material model.  Updates include a switch option to use the Drucker Prager or stress limiter option and documentation for the stress limiter option.

### CODE REVIEW

File: include/aspect/material_model/visco_plastic.h [hunk 1]
```diff
@@ -162,7 +162,15 @@ namespace aspect
           composite
         } viscous_flow_law;
 
-
+        /**
+         * Enumeration for selecting which type of yield mechanism to use.
+         * Select between drucker prager and stress limiter.
+         */
+        enum yield_scheme
```
Comment (bangerth) [inline include/aspect/material_model/visco_plastic.h]: Can you use CameCase for types? So here `YieldScheme` (or, if you want to match the name of the variable, `YieldMechanism`.

File: include/aspect/material_model/visco_plastic.h [hunk 2]
```diff
@@ -173,7 +181,8 @@ namespace aspect
                                           const double &pressure,
                                           const double &temperature,
                                           const SymmetricTensor<2,dim> &strain_rate,
-                                          const enum viscosity_scheme &viscous_type) const;
+                                          const enum viscosity_scheme &viscous_type,
+                                          const enum yield_scheme &yield_type) const;
```
Comment (bangerth) [inline include/aspect/material_model/visco_plastic.h]: C++ doesn't require to say `enum` here. Just omit.

File: include/aspect/material_model/visco_plastic.h [hunk 3]
```diff
@@ -190,6 +199,8 @@ namespace aspect
         std::vector<double> angles_internal_friction;
         std::vector<double> cohesions;
 
+        std::vector<double> exponents_stress_limiter;
+
```
Comment (bangerth) [inline include/aspect/material_model/visco_plastic.h]: no need for the empty line

File: source/material_model/visco_plastic.cc [hunk 4]
```diff
@@ -132,7 +132,8 @@ namespace aspect
                                       const double &pressure,
                                       const double &temperature,
                                       const SymmetricTensor<2,dim> &strain_rate,
-                                      const enum viscosity_scheme &viscous_type) const
+                                      const enum viscosity_scheme &viscous_type,
+                                      const enum yield_scheme &yield_type) const
```
Comment (bangerth) [inline source/material_model/visco_plastic.cc]: no `enum` needed here

File: source/material_model/visco_plastic.cc [hunk 5]
```diff
@@ -395,7 +425,10 @@ namespace aspect
                              "Select what type of viscosity law to use between diffusion, "
                              "dislocation and composite options. Soon there will be an option "
                              "to select a specific flow law for each assigned composition ");
-
+          prm.declare_entry ("Yield mechanism", "drucker",
+                             Patterns::Selection("drucker|limiter"),
+                             "Select what type of yield mechanism to use between drucker prager "
```
Comment (bangerth) [inline source/material_model/visco_plastic.cc]: maybe capitalize `drucker prager`?

File: source/material_model/visco_plastic.cc [hunk 6]
```diff
@@ -395,7 +425,10 @@ namespace aspect
...
+                             "and stress limiter options.");
```
Comment (bangerth) [inline source/material_model/visco_plastic.cc]: I suppose these mechanisms are documented in the plugin documentation?

File: source/material_model/visco_plastic.cc [hunk 7]
```diff
@@ -463,6 +496,13 @@ namespace aspect
                              "The extremely large default cohesion value (1e20 Pa) prevents the viscous stress from "
                              "exceeding the yield stress. Units: $Pa$.");
 
+          // Stress limiter parameters
+          prm.declare_entry ("Stress limiter exponents", "1.0",
+                             Patterns::List(Patterns::Double(0)),
+                             "List of stress limiter exponents, $n_lim$, for background material and compositional fields, "
```
Comment (bangerth) [inline source/material_model/visco_plastic.cc]: `n_\\text{lim}`

Comment (bangerth) [inline source/material_model/visco_plastic.cc]: The line is too long to see it on github, which likely means that it's also generally too long, I suppose :-)

File: source/material_model/visco_plastic.cc [hunk 8]
```diff
@@ -524,6 +564,14 @@ namespace aspect
             AssertThrow(false, ExcMessage("Not a valid viscous flow law"));
 
           // Rheological parameters
+          if (prm.get ("Yield mechanism") == "drucker")
+            yield_mechanism = drucker_prager;
+          else if (prm.get ("Yield mechanism") == "limiter")
+            yield_mechanism = stress_limiter;
+          else
+            AssertThrow(false, ExcMessage("Not a valid yield mechanism"));
```
Comment (bangerth) [inline source/material_model/visco_plastic.cc]: put a period at the end of the error text

File: source/material_model/visco_plastic.cc [hunk 9]
```diff
@@ -600,7 +653,26 @@ namespace aspect
                                    "the yield stress, the viscosity is rescaled back to the yield "
                                    "surface: $v_{y}=\\sigma_{y}/(2.*{\\varepsilon}_{ii})$. "
                                    "This form of plasticity is commonly used in geodynamic models "
-                                   "See, for example, Thieulot, C. (2011), PEPI 188, 47-68. "
+                                   "See, for example, Thieulot, C. (2011), PEPI 188, p. 47-68. "
```
Comment (bangerth) [inline source/material_model/visco_plastic.cc]: p. -> pp.

File: source/material_model/visco_plastic.cc [hunk 10]
```diff
@@ -600,7 +653,26 @@ namespace aspect
...
+                                   "\n"
```
Comment (bangerth) [inline source/material_model/visco_plastic.cc]: For weird reasons, you have to use `\n\n` here

File: source/material_model/visco_plastic.cc [hunk 11]
```diff
@@ -600,7 +653,26 @@ namespace aspect
...
+                                   "Viscous stress may also be limited by a non-linear stress limiter "
+                                   "that has a form similar to the peierls creep mechanism. "
```
Comment (bangerth) [inline source/material_model/visco_plastic.cc]: capitalize Peierls

File: source/material_model/visco_plastic.cc [hunk 12]
```diff
@@ -600,7 +653,26 @@ namespace aspect
...
+                                   "reference strain rate, $\\varepsilon_ii$ is the strain rate "
+                                   "and $n_y$ is the stress limiter exponent.  The yield stress, "
+                                   "$\\tau_y$, is defined through the Drucker Prager yield criterion "
+                                   "formulation.  This method of limiting viscous stress has been used "
+                                   "in various forms within the geodynamic literature, including "
+                                   "Christensen (1992), JGR, 97(B2), p. 2015-2036; "
+                                   "Cizkova and Bina (2013), EPSL, 379, p. 95-103; "
+                                   "Cizkova and Bina (2015), EPSL, 430, p. 408-415. "
+                                   "When $n_y$ is 1, it essentially becomes a linear viscosity model,"
+                                   "and in the limit $n_y\rightarrow \infty$ it converges to the "
```
Comment (bangerth) [inline source/material_model/visco_plastic.cc]: -> `\\infty`

File: source/material_model/visco_plastic.cc [hunk 13]
```diff
@@ -600,7 +653,26 @@ namespace aspect
...
+                                   "{\\frac{\\varepsilon_ii}{\\varepislon_y}}^{\\frac{1}{n_y}-1}$ "
+                                   "Above $tau_y$ is a yield stress, $varepsilon_y$ is the "
+                                   "reference strain rate, $\\varepsilon_ii$ is the strain rate "
+                                   "and $n_y$ is the stress limiter exponent.  The yield stress, "
+                                   "$\\tau_y$, is defined through the Drucker Prager yield criterion "
+                                   "formulation.  This method of limiting viscous stress has been used "
+                                   "in various forms within the geodynamic literature, including "
+                                   "Christensen (1992), JGR, 97(B2), p. 2015-2036; "
+                                   "Cizkova and Bina (2013), EPSL, 379, p. 95-103; "
+                                   "Cizkova and Bina (2015), EPSL, 430, p. 408-415. "
```
Comment (bangerth) [inline source/material_model/visco_plastic.cc]: all three lines: p. -> pp.

File: source/material_model/visco_plastic.cc [hunk 14]
```diff
@@ -600,7 +653,26 @@ namespace aspect
...
                                    "This form of plasticity is commonly used in geodynamic models "
-                                   "See, for example, Thieulot, C. (2011), PEPI 188, 47-68. "
+                                   "See, for example, Thieulot, C. (2011), PEPI 188, p. 47-68. "
+                                   "\n"
+                                   "Viscous stress may also be limited by a non-linear stress limiter "
+                                   "that has a form similar to the peierls creep mechanism. "
+                                   "This stress limiter assigns an effective viscosity "
+                                   "$\\sigma_eff = \\frac{\\tau_y}{2*\\varepsilon_y} "
+                                   "{\\frac{\\varepsilon_ii}{\\varepislon_y}}^{\\frac{1}{n_y}-1}$ "
+                                   "Above $tau_y$ is a yield stress, $varepsilon_y$ is the "
```
Comment (bangerth) [inline source/material_model/visco_plastic.cc]: double backslashes missing on tau and varepsilon

File: source/material_model/visco_plastic.cc [hunk 15]
```diff
@@ -600,7 +653,26 @@ namespace aspect
...
+                                   "reference strain rate, $\\varepsilon_ii$ is the strain rate "
```
Comment (bangerth) [inline source/material_model/visco_plastic.cc]: you need braces around `ii`

**Comment by bangerth** [comment]
As discussed, a new test would be great!

File: source/material_model/visco_plastic.cc (hunk 6 above)
Comment (naliboff) [inline source/material_model/visco_plastic.cc]: Indeed

**Comment by bangerth** [comment]
What's the status of this right now? There's only one unaddressed comment left, it seems, plus the addition of a test. 

Might be useful to squash everything into one commit at the end.

**Comment by naliboff** [comment]
Fixing the last suggestion right now and adding the test.  Will then squash into a single commit.

**Pull request merged**